Public/Src/Sandbox/Windows/DetoursServices/DetouredFunctions.cpp

```diff
@@ -1778,6 +1778,224 @@ NTSTATUS NTAPI Detoured_ZwSetInformationFile(
         FileInformationClass);
 }
 
+/// Runs an injected AnyBuild shim instead of a child process, passing the
+/// original command and arguments to the shim along with, implicitly,
+/// the current working directory and environment.
+BOOL WINAPI InjectShim(
+    _In_opt_    LPCWSTR               lpApplicationName,
+    _Inout_opt_ LPWSTR                lpCommandLine,
+    _In_opt_    LPSECURITY_ATTRIBUTES lpProcessAttributes,
+    _In_opt_    LPSECURITY_ATTRIBUTES lpThreadAttributes,
+    _In_        BOOL                  bInheritHandles,
+    _In_        DWORD                 dwCreationFlags,
+    _In_opt_    LPVOID                lpEnvironment,
+    _In_opt_    LPCWSTR               lpCurrentDirectory,
+    _In_        LPSTARTUPINFOW        lpStartupInfo,
+    _Out_       LPPROCESS_INFORMATION lpProcessInformation)
+{
+    // Create a final buffer for the original command line.
+    size_t fullCmdLineSizeInChars =
+        (lpApplicationName != nullptr ? wcslen(lpApplicationName) : 0) +
+        (lpCommandLine != nullptr ? wcslen(lpCommandLine) : 0) +
+        4;  // Quotes and space and trailing null
+    wchar_t *fullCommandLine = new wchar_t[fullCmdLineSizeInChars];
+    if (fullCommandLine == nullptr)
+    {
+        Dbg(L"Failure running AnyBuild shim process - failed to allocate buffer of size %d.", fullCmdLineSizeInChars * sizeof(WCHAR));
+        SetLastError(ERROR_OUTOFMEMORY);
+        return FALSE;
+    }
+
+    fullCommandLine[0] = L'\0';
+    if (lpApplicationName != nullptr)
+    {
+        fullCommandLine[0] = L'"';

```

Comment:
>ne[0] = L [](start = 21, length = 9)

shouldn't we only do this if there is a space?
Some tools actually print the entire commandline and now this is changing behavior

Reply:
The more I think about it is that it seems the shim tool will need separation between the cmd and args fields and don't try to reconstrcuct a full commandline. You need it for the matching but when executing it would probably be safest to maintain compability

---
In reply to: [282958577](https://github.com/microsoft/BuildXL/pull/279#discussion_r282958577) [](ancestors = 282958577)

Reply:
The logic does in fact do the command pull-out for analysis but in the end when shimming is replaces the lpCommandName with the shim, and prepends the lpCommandLine with the original command (if not null) surrounded in quotes, to make it easier to parse in the shim.